core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+use {
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        collections::{hash_map::Entry, HashMap},
+        fmt::{Debug, Display},
+        ops::{BitAndAssign, Sub},
+    },
+};
+
+pub const MAX_THREADS: usize = u64::BITS as usize;
+
+/// Identifier for a thread
+pub type ThreadId = usize; // 0..MAX_THREADS-1
+
+/// A bit-set of threads an account is scheduled or can be scheduled for.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ThreadSet(u64);
+
+/// Thread-aware account locks which allows for scheduling on threads
+/// that already hold locks on the account. This is useful for allowing
+/// queued transactions to be scheduled on a thread while the transaction
+/// is still being executed on the thread, up to a queue limit.
+pub struct ThreadAwareAccountLocks {
+    /// Number of threads.
+    num_threads: usize, // 0..MAX_THREADS
```

Comment:
bonus: `num_threads` and `sequential_queue_limit` can be promoted to const generics. (const `num_threads` can further propagated down into `ThreadSet` as well).

Reply:
num_threads cannot be known at compile time, since this is the number of threads we have available to schedule on, which is controlled by an env var.

I could probably make the sequential limit a const generic, but want to leave that open for now since we may want to allow a tweak of that via env var as well. Thoughts I had here were that on slower machines which are not able to process txs as fast, we want to queue less work so that we're more likely to get higher priority txs scheduled if they come in later. Faster machines can queue a bit more since they will be able to process all before the higher prio tx comes in.

Actually, now that I've explained that. I wonder if this is misplaced, and the sequential queue limit is better placed within the scheduler logic. What do you think about that?